### ISSUE TITLE
As a user, I want to add a new recurring task so that I can prevent the need of keying the same cyclical task every period

### ISSUE THREAD

**Issue by louislai** [CS2103JAN2017-W13-B3/main#76]

**Comment by Melvin-Tan** [comment]
Initial idea before I execute them
So I'll need extra columns for Task
- recursUntil: datetime
- frequency: [daily, weekly, monthly, yearly]
Validates: Must have at least endDate (i.e. cannot be a recurring floating task)

Add:
"add task by/friday 9pm repeat/weekly until/next year"

Update:
Can update recurringuntil and frequency as usual.
If user tries to update non-recurring tasks to recurring, they MUST provide both frequency (recurringuntil is optional), or we throw error.
If user tries to update recurring task to non-recurring NON-FLOATING task (e.g. recurring event to nonrecurring task with deadline), they can just do "stoprepeating/" to set them back to NULL.
If user tries to update recurring task to non-recurring FLOATING task, simply use the "floating/" in the command.

Delete 1 recurrence:
No need to change anything. Just delete 1.
Delete whole task:
Update to remove recurrence first using stoprepeating/, then delete this task by using delete 1.

Key operations:
Everytime user marks task as complete OR when user delete this task, we generate the next one for them.

Also, CS2103T handbook seems to suggest need to support EXCEPTION to the rule, like...
"CS2103 lectures recurring from week 1 to week 13 except in recess week."
LETS NOT DO THIS.

**Comment by louislai** [comment]
@Rinder5

**Comment by louislai** [comment]
let's not have exception.

**Comment by louislai** [comment]
So if you I understand correctly, `"add task by/friday 9pm recurringuntil/next year frequency/daily" will means tasks recur everyday at 9pm?

```
Delete 1 recurrence:
No need to change anything.
Delete whole task:
Update to remove recurrence first, then delete this task.
```

So deleting a task wont actually delete the task but only 1 occurence? How exactly will that look?

`(The alternative plan is I just generate all of them until endDate, but this will clutter the view or cause too many tasks to render at once).` might cause issues imo. If you update the recurring frequency you would then need update multiple tasks at once, and possibly remove/add tasks

**Comment by louislai** [comment]
best way to handle it is probably not to invent our own approach but follow a well-established approach. Microsoft Project supports recurring tasks as well, so can take it as reference on how they do it https://support.office.com/en-us/article/Create-recurring-tasks-88a4d903-38c3-4665-870e-4810b752f2c4

Another example http://www.michaellinenberger.com/blog/using-recurring-tasks-in-outlook/
``` future instances of recurring tasks are not placed in your task list until the previous one is marked complete or deleted. So they are a convenient way to keep a repeating task on track without burdening your task list with a long list of future tasks. ``` this is the more sane option imho

**Comment by Rinder5** [comment]
ditto for not having exception

**Comment by Rinder5** [comment]
```
Delete 1 recurrence:
No need to change anything.
Delete whole task:
Update to remove recurrence first, then delete this task.
```
for your deletion, how do you plan to differentiate user command for delete 1 vs delete recursion?

```
So what I have in mind is that everytime an endDate is over OR when user delete this task, we generate the next one for them.
```
do you mean when user completes the task, not deletes? if so, we will need to rethink about naming convention or remove unique naming. if not, the user will have to use delete instead of mark for this type of task, which will be confusing for them.

I think your first section is good though.
Any plans for recurring events? e.g. thurs-friday recurring weekly

**Comment by louislai** [comment]
interestingly our school outlook can act as a task manager, and they have recurring tasks. They allow option for deleting single occurence or all occurrences. For us, we can do it like
- `delete 1` to delete all occurences (for consistency)
- `delete 1 single` to delete only 1 occurence

Or just keep deleting all to maintain @Melvin-Tan 's sanity. There are ways to make this feature more complicated, but I rlly dont wanna go there

**Comment by Melvin-Tan** [comment]
Hmm sorry i'll elaborate on this delete 1 vs delete all recurrence
Suppose i have a weekly recurring task at index 1.
If i want to remove just this week, i can just do delete 1, then it gets removed and the next week's gets generated.

If i want to remove the whole thing entirely, i must do something like making it non-recurring first (update 1 recurringuntil/ frequency/, or a custom one like update 1 stoprecurring/), then delete 1.

**Comment by Melvin-Tan** [comment]
@Rinder5 recurring events will be treated the exact same way as recurring task with deadline
Except it has start and end date instead of just end date.

**Comment by Melvin-Tan** [comment]
@louislai 
So if you I understand correctly, `"add task by/friday 9pm recurringuntil/next year frequency/daily" will means tasks recur everyday at 9pm?

Correct. Theoretically speaking.......
Suppose the friday has a date of 12 Aug 2018 (anyhow one)
If frequency is weekly, then it means the task is a task with deadline every friday 9pm.
If frequency is month, then it means the task is a task with deadline every 12th of the month, 9pm.
If frequency is yearly, then it means the task is a task with deadline every 12 Aug, 9pm.

**Comment by Melvin-Tan** [comment]
@Rinder5 
`So what I have in mind is that everytime an endDate is over OR when user delete this task, we generate the next one for them.`
do you mean when user completes the task, not deletes? if so, we will need to rethink about naming convention or remove unique naming. if not, the user will have to use delete instead of mark for this type of task, which will be confusing for them.

I'll probably go with Louis approach of "future instances of recurring tasks are not placed in your task list until the previous one is marked complete or deleted." + when endDate is over.
Either delete or mark can trigger the next task creation, so it is fine i think. This is also consistent with my "delete 1" deleting only 1 instance of the recurring.

**Comment by Melvin-Tan** [comment]
Another issue I just thought of, 
should I enforce recurringUntil parameter?
i.e. if they dont provide a stop date, should i let it recur forever?
If no, that makes things simpler. If yes, what would be the best approach for recur forever? Putting a date 1000 years later for them?

**Comment by Rinder5** [comment]
```
"future instances of recurring tasks are not placed in your task list until the previous one is marked complete or deleted."
```
So a new task will be created with the same name? If you still trigger this by mark complete, we will have 2 similar named tasks.

```
So if you I understand correctly, `"add task by/friday 9pm recurringuntil/next year frequency/daily" will means tasks recur everyday at 9pm?

Correct. Theoretically speaking.......
```
I think the point that @louislai is making here is that for example, we will add a task with deadline friday 9pm each day. With the way your key operation works, this means that it will only update weekly, correct? Since the deadline is not over, it will not create a new one. If and only if the user completes it, then it will add it back the next day?

**Comment by Rinder5** [comment]
recurringUntil should not be enforced. e.g. for listing habits (brush teeth/eat breakfast/etc.)

You can implement it however you want, but on the user side it should be optional.

**Comment by Melvin-Tan** [comment]
Same task same name with different start/end dates is still considered a different task. That shouldn't be a problem. Yes the new task can be created with the same name.

No, if you state daily, i am going to ignore the day and just create one every 9pm. You shouldnt (and probably have no use) try to add task with deadline friday 9pm every single day, or u will get 6 duplicates of the same thing.
I am not sure if I am clear enough on this sorry for the bad explanation lol.

**Comment by Rinder5** [comment]
so if the user enters that you will throw error? if so, then everything looks fine

**Comment by Melvin-Tan** [comment]
Hmmm no need to throw error, cuz they can say like... by/24 Mar 2018 9pm (which we also know is a Friday, but we are just gonna ignore the Friday portion and just do it every 9pm)
If they don't give a timing and say daily, we slap a default timing for them.

**Comment by louislai** [comment]
Resolved by #149
